controller/backing_image_controller.go

```diff
@@ -291,9 +304,133 @@ func (bic *BackingImageController) syncBackingImage(key string) (err error) {
+		if err := bic.handleBackingImageCopiesEvictions(nonFailedCopies, bi); err != nil {
+			return nil
+		}
+	} else { // nonFailedCopies < bi.Spec.MinNumberOfCopies
+		readyNode, readyDiskName, err := bic.ds.GetReadyNodeDiskForBackingImage(bi)
+		logrus.Infof("replicate the copy to node: %v, disk: %v", readyNode, readyDiskName)
+		if err != nil {
+			logrus.WithError(err).Warnf("failed to create the backing image copy")
+			return nil
+		}
+		// BackingImageManager will then sync the BackingImage to the disk
+		bi.Spec.DiskFileSpecMap[readyNode.Status.DiskStatus[readyDiskName].DiskUUID] = &longhorn.BackingImageDiskFileSpec{}
+	}
+
 	return nil
 }
 
+// handleBackingImageCopiesEvictions do creating one more replica for eviction, if requested
+func (bic *BackingImageController) handleBackingImageCopiesEvictions(nonFailedCopies int, bi *longhorn.BackingImage) (err error) {
+	log := getLoggerForBackingImage(bic.logger, bi)
+	NonEvictingCount := nonFailedCopies
+
+	for _, fileSpec := range bi.Spec.DiskFileSpecMap {
+		if fileSpec.EvictionRequested {
+			NonEvictingCount--
+		}
+	}
+
+	if NonEvictingCount < bi.Spec.MinNumberOfCopies {
+		log.Infof("Creating one more backing image copy for eviction")
+		readyNode, readyDiskName, err := bic.ds.GetReadyNodeDiskForBackingImage(bi)
+		if err != nil {
+			logrus.WithError(err).Warnf("failed to create the backing image copy")
+			return nil
+		}
+		// BackingImageManager will then sync the BackingImage to the disk
+		bi.Spec.DiskFileSpecMap[readyNode.Status.DiskStatus[readyDiskName].DiskUUID] = &longhorn.BackingImageDiskFileSpec{}
+	}
+
+	return nil
+}
+
+func (bic *BackingImageController) cleanupEvictionRequestedBackingImageCopies(bi *longhorn.BackingImage) {
+	log := getLoggerForBackingImage(bic.logger, bi)
+
+	// If there is no non-evicting healthy backing image copy,
+	// Longhorn should retain one evicting healthy backing image copy for replenishing.
+	hasNonEvictingHealthyBackingImageCopy := false
+	evictingHealthyBackingImageCopyDiskUUID := ""
+	for diskUUID, fileSpec := range bi.Spec.DiskFileSpecMap {
+		fileStatus, exists := bi.Status.DiskFileStatusMap[diskUUID]
+		if exists && fileStatus != nil {
+			if fileStatus.State != longhorn.BackingImageStateReady {
+				continue
+			}
+		}
+
+		if !fileSpec.EvictionRequested {
+			hasNonEvictingHealthyBackingImageCopy = true
+			break
+		}
+		evictingHealthyBackingImageCopyDiskUUID = diskUUID
+	}
+
+	for diskUUID, fileSpec := range bi.Spec.DiskFileSpecMap {
+		if !fileSpec.EvictionRequested {
+			continue
+		}
+		if !hasNonEvictingHealthyBackingImageCopy && diskUUID == evictingHealthyBackingImageCopyDiskUUID {
+			msg := fmt.Sprintf("Failed to evict backing image copy on disk %v for now since there is no other healthy backing image copy", diskUUID)
+			log.Warn(msg)
+			bic.eventRecorder.Event(bi, corev1.EventTypeNormal, constant.EventReasonFailedDeleting, msg)
+			continue
+		}
+		delete(bi.Spec.DiskFileSpecMap, diskUUID)
```

Comment:
Wait... Isn't `NodeController.syncBackingImageEvictionRequested` responsible for this?

In general, the resource controller cannot update its CR spec.

Reply:
sure, I think we can move the logic to the node controller.

Reply:
Hi @shuo-wu 
I think keep the delete copy logic in the backing image controller is better.

> Isn't NodeController.syncBackingImageEvictionRequested responsible for this?
inside the NodeController, it only update the eviction request to true for the copy on the node or disks
So the BackingImageController can then replenish the copy to other nodes/disks

> In general, the resource controller cannot update its CR spec.

I know the reason is to prevent unlimited updating loop. But
- we actually already put the **replenish copy logic in the BackingImageController** as well which will add the disk to the `bi.Spec`.
- It is much better to put it here then the node controller. Here is the reason
     - to decide if we can evict the copy on one disk needs the knowledge of all the copies of the backing image
     - NodeController only consider its own node and disks. But backing image copies can be on two nodes
     - It could possible that two node controller all decide to delete one of the evicted the backing image on their node at the same time based on the status it gets at the moment. Thus, the race condition can happen, perhaps that is the last healthy one because the other NodeController already delete the other one.

Put the eviction logic and replenish logic in the BackingImageController is better.
Because it handle the backing image one at the time and has the whole picture of the status.
Thus it actually won't get into unlimited update loop because it will get balanced on the minimal HA number in limited reconcile times.

Reply:
Got it, the explanation makes sense. 

Adding or deleting entries for `bi.Spec.DiskFileSpecMap` involves multiple nodes. Asking a specific NodeController to check other nodes for that field is not a good idea.

In other words, the BackingImage is similar to Volume CRD while file copies of a BackingImage are similar to replicas of a volume. By contrast, it's BackingImageController's responsibility to handle the BI copies. Currently, I don't want to introduce a new CRD for the copies hence the controller updating its own CR spec looks weird but somehow is reasonable.

@derekbit Are you fine with breaking the controller convention here?

Reply:
>  I don't want to introduce a new CRD for the copies hence the controller updating its own CR spec looks weird but somehow is reasonable.

A bit weird but I was struggling with it as well.
To avoid potential deadlock issues, it would be better to move this to another controller. However, I'm fine with the current implementation and we can optimize it later if it requires significant effort.


> the BackingImage is similar to Volume CRD while file copies of a BackingImage are similar to replicas of a volume.

Yes, we might need a high-level CRD (like a volume) for `BackingImage` (like a replica) in the future.

Reply:
As discussed, for current version we still implement the logic in the BackingImageController
We will refactor this part in the future with this issue: https://github.com/longhorn/longhorn/issues/8783